### PR TITLE
fix: use effective bodyweight load for exercise display and 1RM (#684)

### LINKED ISSUE #684: Incorrect weight in bodyweight exercise Personal Records (Decline Push Up)
## Bug Report: Incorrect Weight in Bodyweight Exercise Personal Records

**Reporter:** stugotz0955_50891 (Discord)
**Type:** Bug
**Repo:** Mobile (project-phoenix-mp)
**Platform:** Mobile app

### Summary

Incorrect weight is being graphed and displayed in Personal Records for the Decline Push Up bodyweight exercise. The displayed weight does not appear to use the calculated weight based on the user's bodyweight, additional weight from a weighted vest, and the percentage for movement and decline height. Volume and reps are displayed correctly.

### Steps to Reproduce

1. Set up bodyweight exercise configuration with bodyweight, weighted vest additional weight, and movement/decline percentage
2. Perform Decline Push Up exercises and log sets
3. Navigate to Personal Records for Decline Push Up
4. Observe the weight displayed in the PR graph/data

### Expected Behavior

The Personal Records weight for bodyweight exercises should use the calculated composite weight: bodyweight + additional weight (weighted vest) × percentage for movement type and decline height. This matches the formula used during the workout itself.

### Actual Behavior

The weight displayed and graphed in Personal Records is incorrect — it does not incorporate the full calculated weight formula (bodyweight + vest weight + movement/decline percentage). Volume and reps are correct, suggesting the issue is isolated to the weight calculation or display path in the PR system.

### Impact

Users who use weighted vests or bodyweight percentage calculations for exercises like Decline Push Up see inaccurate historical weight data in their Personal Records, making it impossible to track true strength progression for bodyweight exercises with additional load.

### Attachments

- Screenshot: [get.png](https://cdn.discordapp.com/attachments/1509574296807407717/1530613980643725352/get.png?ex=6a6636da&is=6a64e55a&hm=a08ef8332438372af4f5a088b361e65d172ed70e87beba5ff8dcac32beacd0bc&) — shows the incorrect weight display in the PR screen

### Original Discord Report

```
<@1502654170031849644> Incorrect weight being graphed and displayed in bodyweight exercise Personal Records for Decline Push Up.  It doesn't appear to be using the calculated weight based on my bodyweight, any additional weight from the weighted vest used and percentage for movement and decline height.  The volume and reps are correct though.
```

### PR DESCRIPTION
**Pull request by 9thLevelSoftware**
## Summary

Fixes the display-path bug where Exercise Detail / Personal Records shows the configured nominal cable weight instead of the bodyweight-derived effective load for bodyweight exercises (e.g. Decline Push Up).

## Root Cause

`ExerciseDetailScreen.kt` and `ResolveCurrentOneRepMaxUseCase.kt` read `session.weightPerCableKg` (the configured/nominal machine load) in 4 display paths. For bodyweight exercises, the correct effective load is persisted in `session.heaviestLiftKg` during the workout, and `effectiveHeaviestKgPerCable()` already provides the backward-compatible accessor (`heaviestLiftKg ?: weightPerCableKg`).

Volume was already correct because it uses `effectiveTotalVolumeKg()` — this fix aligns the weight and 1RM display paths with the same pattern.

## Changes

| File | Change |
|---|---|
| `ExerciseDetailScreen.kt` | weightTrendData, ExerciseHistoryTable weight column, SessionHistoryRow: `weightPerCableKg` → `effectiveHeaviestKgPerCable()` |
| `ResolveCurrentOneRepMaxUseCase.kt` | `estimatedOneRepMaxPerCableOrNull()`: same switch |
| `ResolveCurrentOneRepMaxUseCaseTest.kt` | New regression test for bodyweight sessions with effective load, weighted-vest sessions, and legacy fallback |

## What this does NOT change

- Bodyweight volume calculation (`BodyweightVolumeCalculator`)
- PR persistence / gamification write path
- Session save contract
- Cable multiplier semantics
- Legacy sessions without `heaviestLiftKg` (they fall through to `weightPerCableKg` via the existing accessor)

## Testing

- New test `session helper uses effectiveHeaviestKgPerCable for bodyweight sessions` covers:
  - Bodyweight session (configured=0, effective=40kg) → 1RM from 40kg
  - Weighted-vest session (configured=5kg, effective=40kg) → 1RM from 40kg
  - Legacy session (no heaviestLiftKg) → 1RM from configured weight
- **Blocker:** No Java Runtime on the CI host; Gradle tests cannot be run locally. CI should verify.

Fixes #684

### CODE REVIEW

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/usecase/ResolveCurrentOneRepMaxUseCase.kt [hunk 1]
```diff
@@ -20,7 +21,7 @@ data class CurrentOneRepMax(
 )
 
 fun WorkoutSession.estimatedOneRepMaxPerCableOrNull(): Float? {
-    val load = weightPerCableKg.takeIf { it.isFinite() && it > 0f } ?: return null
+    val load = effectiveHeaviestKgPerCable().takeIf { it.isFinite() && it > 0f } ?: return null
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/usecase/ResolveCurrentOneRepMaxUseCase.kt]: Fixed in commit 49a31e3: added `weightPerCableKg` fallback when `effectiveHeaviestKgPerCable()` returns 0 (from heaviestLiftKg=0f). This mirrors the SQL `selectExerciseWeightHistory` behavior that treats nonpositive heaviestLiftKg as absent. Added test case covering the edge case.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt [hunk 2]
```diff
@@ -190,8 +191,8 @@ fun ExerciseDetailScreen(
     // Weight-over-time trend data using saved per-cable load.
     val weightTrendData = remember(exerciseSessions) {
         exerciseSessions.mapNotNull { session ->
-            if (session.weightPerCableKg > 0) {
-                session.timestamp to session.weightPerCableKg
+            if (session.effectiveHeaviestKgPerCable() > 0) {
+                session.timestamp to session.effectiveHeaviestKgPerCable()
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt]: Fixed in commit 41f6d84: added displayHeaviestKgPerCable() helper that falls back to weightPerCableKg when effectiveHeaviestKgPerCable() returns 0 (from heaviestLiftKg=0f). Applied to all 3 display paths: weight trend chart, history table weight column, and session history row.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt [hunk 3]
```diff
@@ -187,11 +188,19 @@ fun ExerciseDetailScreen(
     val previousSessionOneRepMax =
         validSessionEstimatesNewestFirst.getOrNull(1)?.second
 
+    // Display-safe heaviest load: fall back to weightPerCableKg when
+    // heaviestLiftKg is 0f (measured zero), matching SQL behavior.
+    fun WorkoutSession.displayHeaviestKgPerCable(): Float =
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt]: Fixed in 092d3cce: moved displayHeaviestKgPerCable to the shared Models.kt domain-model scope, so ExerciseDetailScreen, ExerciseHistoryTable, and SessionHistoryRow all compile against the same implementation. :shared:compileTestKotlinIosArm64 passed.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt [hunk 4]
```diff
@@ -187,11 +188,19 @@ fun ExerciseDetailScreen(
...
+        effectiveHeaviestKgPerCable().takeIf { it > 0f }
+            ?: weightPerCableKg.takeIf { it > 0f }
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/presentation/screen/ExerciseDetailScreen.kt]: Fixed in 092d3cce: displayHeaviestKgPerCable now requires finite measured/configured values and falls back from non-finite summary data to a valid configured load. Added focused coverage for Float.POSITIVE_INFINITY.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/usecase/ResolveCurrentOneRepMaxUseCase.kt [hunk 5]
```diff
@@ -20,7 +21,13 @@ data class CurrentOneRepMax(
...
 
 fun WorkoutSession.estimatedOneRepMaxPerCableOrNull(): Float? {
-    val load = weightPerCableKg.takeIf { it.isFinite() && it > 0f } ?: return null
+    // effectiveHeaviestKgPerCable() returns heaviestLiftKg ?: weightPerCableKg,
+    // but heaviestLiftKg = 0f (measured zero) is treated as "no data" here,
+    // matching the SQL selectExerciseWeightHistory fallback that treats
+    // nonpositive heaviestLiftKg as absent.
+    val load = effectiveHeaviestKgPerCable().takeIf { it.isFinite() && it > 0f }
+        ?: weightPerCableKg.takeIf { it.isFinite() && it > 0f }
+        ?: return null
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/usecase/ResolveCurrentOneRepMaxUseCase.kt]: Fixed in 092d3cce: an explicit zero heaviestLiftKg with persisted positive counterweight is preserved as a real zero effective bodyweight load; the history display shows zero and 1RM returns null instead of using the unrelated nominal machine load. Added regression coverage.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/model/Models.kt [hunk 6]
```diff
@@ -608,6 +608,24 @@ data class WorkoutSession(
...
+ * preserved only when persisted counterweight context proves that a bodyweight load was
+ * genuinely offset to zero; otherwise zero, negative, and non-finite summary values fall
+ * back to the finite positive configured value. Legacy sessions have no summary value and
+ * therefore use their configured weight unchanged.
+ */
+fun WorkoutSession.displayHeaviestKgPerCable(): Float {
+    val measured = heaviestLiftKg
+    if (measured != null && measured.isFinite()) {
+        if (measured > 0f) return measured
+        if (measured == 0f && counterweightKg.isFinite() && counterweightKg > 0f) return 0f
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/model/Models.kt]: Fixed in a27d32db: removed the counterweight-only zero preservation. `WorkoutSession` persists neither `Exercise.isBodyweight` nor an equivalent classification, and `captureRackLoadSnapshot` records counterweights for machine exercises too, so the persisted fields cannot prove that zero is intentional. `displayHeaviestKgPerCable()` now uses a finite positive measured value only; every zero/negative/non-finite summary falls back to finite positive `weightPerCableKg`, matching the existing SQL history contract. Added assertions for both display and session-based 1RM behavior with zero summary + counterweight.

File: shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/model/Models.kt [hunk 7]
```diff
@@ -608,6 +608,21 @@ data class WorkoutSession(
...
+ * A finite positive recorded summary value is authoritative. Zero, negative, and non-finite
+ * summary values fall back to the finite positive configured value. A WorkoutSession does not
+ * persist the exercise bodyweight classification, so counterweight alone cannot safely identify
+ * an intentional zero bodyweight load. Legacy sessions have no summary value and therefore use
+ * their configured weight unchanged.
+ */
+fun WorkoutSession.displayHeaviestKgPerCable(): Float {
+    val measured = heaviestLiftKg
+    if (measured != null && measured.isFinite() && measured > 0f) return measured
+    return weightPerCableKg.takeIf { it.isFinite() && it > 0f } ?: 0f
```
[the remark this answers is not in the document]
Comment (9thLevelSoftware) [reply shared/src/commonMain/kotlin/com/devil/phoenixproject/domain/model/Models.kt:634]: Fixed in 766c25e1. The Exercise Detail screen now fetches the canonical `Exercise.isBodyweight` value, carries it through the request/load gate into `ResolveCurrentOneRepMaxUseCase`, and passes it to every chart/table/history resolution. A zero is preserved only for `isBodyweight=true` plus positive persisted counterweight; the default and explicit non-bodyweight cases retain the SQL configured-load fallback. Added focused model/1RM assertions and a load-gate test proving the context reaches the resolver.
